1.13.0-rc4 cherry-picks: part2

vieux opened the pull request:
#Fix `docker exec -u` issue after docker daemon restart #29365
[plugins] return err when failing remove #29372
Add the link for service logs #29383
Add docs for `docker network create --attachable` #29412
Make bash completion for `docker stack deploy --bundle-file` available only in experimental mode #29413
fix typo in reference for service create #29419 
Show usage when `docker swarm update` has no flags #29000
Make graphdriver plugin use plugin BasePath #29302
fix devices in plugins #29462
swarm leave is not only for workers #29482
Add the whole related information for node #29474
Fix inconsistency for pause and unpause #29434
Remove fedora-23 because eol #29455
api: allow NW name that is the prefix of a swarm NW ID #27938
Fix volume plugin refcounting on daemon restart #29042

AkihiroSuda committed:
api: allow creating a network of which name is the prefix of the ID of a swarm network

Previously, it doesn't allow creating such a network:

e.g.

    $ docker network inspect -f '{{.Id}}' ingress
    84xh9knigj6zyt00u31e26nj3
    $ docker network create 84
    Error response from daemon: network with name 84 already exists

Fix #27866

Signed-off-by: Akihiro Suda <suda.akihiro@lab.ntt.co.jp>
(cherry picked from commit edfbc3b8767ab2e89e73ba3142d2ddad295001e9)
Signed-off-by: Victor Vieux <vieux@docker.com>

yongtang committed:
Show usage when `docker swarm update` has no flags

This fix tries to address the issue raised in 24352. Previously,
when `docker swarm update` has no flags, the output is
```
Swarm updated.
```
even though nothing was updated. This could be misleading for
users.

This fix tries to address the issue by adding a `PreRunE` function
in the command so that in case no flag is provided (`cmd.Flags().NFlag() == 0`),
the usage will be outputed instead.

An integration has been added to cover the changes.

This fix fixes 24352.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>
(cherry picked from commit 5aa5a1cb008963ee6a99011b8f7cd1f489e2bc6a)
Signed-off-by: Victor Vieux <vieux@docker.com>

cpuguy83 committed:
Fix volume plugin refecounting on daemon restart

Ensures all known volumes (known b/c they are persisted to disk) have
their volume drivers refcounted properly.

In testing this, I found an issue with `--live-restore` (required since
currently the provided volume plugin doesn't keep state on restart)
where restorted plugins did not have a plugin client loaded causing a
panic when trying to use the plugin.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>
(cherry picked from commit 6ef1060cd0acb847e06db890abb335faa837a9e2)
Signed-off-by: Victor Vieux <vieux@docker.com>

cpuguy83 committed:
Make graphdriver plugin use plugin BasePath

Also enables `PropagatedMount` for graphdrivers.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>
(cherry picked from commit 500210475f6d841b2eacb42fb495e90108db2733)
Signed-off-by: Victor Vieux <vieux@docker.com>

tophj-ibm committed:
[plugins] return err when failing remove

Fixes a case where removing the plugin from disk would
fail silently. Also moves pluginStore remove after we
remove from disk, so 'docker plugin ls' doesn't isn't
empty in case it errors out.

Signed-off-by: Christopher Jones <tophj@linux.vnet.ibm.com>
(cherry picked from commit fb11164c4f71315c96ddf3de46ff4689a5bac59c)
Signed-off-by: Victor Vieux <vieux@docker.com>

yuexiao-wang committed:
Add the whole related information for node

Signed-off-by: yuexiao-wang <wang.yuexiao@zte.com.cn>
(cherry picked from commit b67b0949ed4b528fef6c3dc059a4bd16d5f2a929)
Signed-off-by: Victor Vieux <vieux@docker.com>

yongtang committed:
Fix `docker exec -u` issue after docker daemon restart

This fix tries to address the issue raised in 29342 where
`docker exec -u` after docker daemon restart returns an error:
```
unable to find user test: no matching entries in passwd file
```

The reason was that `container.BaseFS` is not present after restart.

This fix adds the `daemon.Mount` during the restore to bring up the
`container.BaseFS`.

An integration test has been added to cover the changes.

This fix fixes 29342.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>
(cherry picked from commit 7feb2a17e4b9d1a5305a8a44004e916b79cbdd97)
Signed-off-by: Victor Vieux <vieux@docker.com>

yuexiao-wang committed:
Add the link for service logs

Signed-off-by: yuexiao-wang <wang.yuexiao@zte.com.cn>
(cherry picked from commit 54d38fe115e0183af7d862e54ba6e58c347e2a8d)
Signed-off-by: Victor Vieux <vieux@docker.com>

albers committed:
Add docs for `docker network create --attachable`

Signed-off-by: Harald Albers <github@albersweb.de>
(cherry picked from commit 7433d3acf61142bb1abbf7928ff0e20acf3eda56)
Signed-off-by: Victor Vieux <vieux@docker.com>

albers committed:
Make bash completion for `docker stack deploy --bundle-file` experimental

Signed-off-by: Harald Albers <github@albersweb.de>
(cherry picked from commit a53bdfa3be6a821b5093919c55f6995e1df35680)
Signed-off-by: Victor Vieux <vieux@docker.com>

Commit:
fix typo in reference for service create

Signed-off-by: Trapier Marshall <trapier.marshall@docker.com>
(cherry picked from commit 90c680ad259057d0065662a1a6fdad779bde6082)
Signed-off-by: Victor Vieux <vieux@docker.com>

yuexiao-wang committed:
Fix inconsistency for pause and unpause

Signed-off-by: yuexiao-wang <wang.yuexiao@zte.com.cn>
(cherry picked from commit 082f4919cac55f32674874d96d82a310f6b1522d)
Signed-off-by: Victor Vieux <vieux@docker.com>

andrewhsu committed:
remove fedora-23 because eol

Signed-off-by: Andrew Hsu <andrewhsu@docker.com>
(cherry picked from commit ef98091a901cd9926d307ed622de088c90da2ef3)
Signed-off-by: Victor Vieux <vieux@docker.com>

vieux committed:
fix devices in plugins

Signed-off-by: Victor Vieux <vieux@docker.com>
(cherry picked from commit 6f00601c9fb7886547334f80a250bac5e226aad7)
Signed-off-by: Victor Vieux <vieux@docker.com>

GordonTheTurtle commented:
Looks like we would not be able to merge this PR because of merge conflicts. Please rebase, fix conflicts, and force push to your branch.

thaJeztah reviewed (approved):
LGTM

thaJeztah committed:
swarm leave is not only for workers

the "docker swarm leave" command description
mentioned that the command can only be used
for workers, however, the command can also
be used for managers (using the `-f` / `--force`
option).

this patch removes the "(workers only)" part
of the command description.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>
(cherry picked from commit f89eee5d9d3c01bc6d498fcb1163ec199e6c3078)
Signed-off-by: Victor Vieux <vieux@docker.com>

vieux commented:
LGTM

tiborvass commented:
LGTM
